libtenzir/builtins/operators/sample.cpp

```diff
@@ -0,0 +1,214 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2024 The Tenzir Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include <tenzir/argument_parser.hpp>
+#include <tenzir/arrow_utils.hpp>
+#include <tenzir/plugin.hpp>
+#include <tenzir/series_builder.hpp>
+#include <tenzir/tql2/plugin.hpp>
+
+#include <arrow/compute/api.h>
+
+namespace tenzir::plugins::sample_ {
+
+namespace {
+
+enum class mode { none, log, log2, log10, sqrt };
+
+class sample_operator final : public crtp_operator<sample_operator> {
+public:
+  sample_operator() = default;
+
+  sample_operator(mode m, duration delta) : mode_{m}, delta_{std::move(delta)} {
+  }
+
+  auto operator()(generator<table_slice> input,
+                  operator_control_plane&) const -> generator<table_slice> {
+    // XXX: Decide if the window should keep rolling even if no events are
+    // received.
+    auto last = std::chrono::steady_clock::now();
+    auto count = uint64_t{};
+    // Logic copied from the slice_operator::positive_stride()
+    // TODO: Consider using the slice operator directly or extracting this
+    // funcitonality.
+    auto offset = int64_t{};
+    auto stride = int64_t{1};
+    const auto calc = [&]() {
+      if (count == 1) {
+        return 1.0;
+      }
+      switch (mode_) {
+        case mode::log:
+          return std::log(count);
+        case mode::log2:
+          return std::log2(count);
+        case mode::log10:
+          return std::log10(count);
+        case mode::sqrt:
+          return std::sqrt(count);
+        case mode::none:
+          TENZIR_UNREACHABLE();
+      }
+      TENZIR_UNREACHABLE();
+    };
+    constexpr auto make_stride_index
+      = [](const int64_t offset, const int64_t rows, const int64_t stride) {
+          TENZIR_ASSERT(stride > 0);
+          auto b = int64_type::make_arrow_builder(arrow::default_memory_pool());
+          check(b->Reserve((rows + 1) / stride));
+          for (auto i = offset % stride; i < rows; i += stride) {
+            check(b->Append(i));
+          }
+          return finish(*b);
+        };
+    for (const auto& slice : input) {
+      if (slice.rows() == 0) {
+        co_yield {};
+        continue;
+      }
+      // XXX: >=?
```

Comment:
I think this doesn't matter as long as you keep it consistent. It really only makes a difference for the first window.